src/analyses/threadId.ml

```diff
@@ -96,15 +87,35 @@ struct
     match x with
     | Queries.CurrentThreadId -> fst ctx.local
     | Queries.CreatedThreads -> created ctx.local
-    | Queries.PartAccess {exp; var_opt; write} ->
-      part_access ctx exp var_opt write
     | Queries.MustBeUniqueThread ->
       begin match fst ctx.local with
         | `Lifted tid -> Thread.is_unique tid
         | _ -> Queries.MustBool.top ()
       end
     | _ -> Queries.Result.top x
 
+  module A =
+  struct
+    (* Also contains MHP in addition to unique thread. *)
+    include Printable.Prod (Printable.Option (ThreadLifted) (struct let name = "nonunique" end)) (MHP)
+    let name () = "thread * mhp"
+    let may_race (t1: t) (t2: t) = match t1, t2 with
+      | (Some t1, _), (Some t2, _) when ThreadLifted.equal t1 t2 -> false
+      | (_, mhp1), (_, mhp2) when not (MHP.may_happen_in_parallel mhp1 mhp2) -> false
+      | (_, _), (_, _) -> true
+    let should_print _ = true
+  end
+  let access ctx e vo w =
+    let unique =
+      if is_unique ctx then
+        let tid = fst ctx.local in
+        Some tid
+      else
+        None
+    in
```

Comment:
Is this actually needed here? Would MHP not also detect this?

Reply:
I wasn't sure, so I kept all the information that we previously had. The logic for this is that a unique thread doesn't race with itself. If MHP implies that, then I can remove this component indeed.

Reply:
https://github.com/goblint/analyzer/blob/f9b403da7ff28458401095f7153bf4391b39b683/src/cdomains/mHP.ml#L29-L35

This should cover that case, right?

Reply:
There is a subtle difference actually. MHP uses just `is_unique` of the thread ID domain, but the one above uses `MustBeUniqueThread` query, which is answered by more analyses. In particular, if simple thread IDs are used and the old thread analysis is activated, it would also work, whereas through the simple thread IDs, MHP cannot determine uniqueness.